text/0114-dynamic-component-configuration.md

```diff
@@ -0,0 +1,233 @@
+import { } from "lds/...";
+
+
+
+export default class LazyCmp extends LightningElement {
+
+@api dashboardId;
+
+@api dashboardData;
+
+@api colorPalette;
+
+@api defaultKpiBreaks;
+
+chartItems = [];
+
+@wire(myAdapter, { id: 'someDashboardId' }) {
+
+	chartDefinitions({ data, error }) {
+
+	chartItems = data.charts.map((chart) => {
+		let ctor;
+		let props = {};
+		let attributes = {};
+		let eventHandlers = {};
+		// Populate 'ctor'
+		// For simplicity assume they've been loaded async elsewhere
+		ctor = chart.ctor;
+		// Populate 'attributes'
+		// For example, change style based on some config data and pass
+		// in a unique id
+		attributes.style = {
+			classname: chart.type === 'donut' ? 'chart-small' : 'chart-medium'
+		};
+		attributes.id = `${dashboardId}-${chart.id}`;
+		// Populate 'props'
+		// Set the dimensions into charts that support them.
+		// For example, a KPI doesn't have a dimensions prop
+		// and would throw if you tried to supply one:
+		if (chart.supportsDimensions) {
+			props.dimensions = dashboardData.dimensions;
+		}
+
+		// Some chart items, like a dimension filter card,
+
+		// might not support measures as inputs:
+
+		if (chart.supportsMeasures) {
+			props.measures = dashboardData.measures
+		}
+
+		// Some charts support colors in different ways:
+		if (chart.type === 'kpi' || chart.type === 'gauge') {
+			props.breakpoints = chart.customBreaks || this.defaultKpiBreaks;
+		} else if (chart.supportsColorPalette) {
+			props.colorPalette = this.colorPalette;
+		}
+		// Populate 'events'
+		if (chart.canFilter) {
+			eventHandlers.onfilter = this.handleFilter;
+		}
+		if (chart.canSort) {
+			eventHandlers.onsort = this.handleSort;
+		}
+		return {
+			ctor,
+			attributes,
+			properties: props,
+			events: eventHandlers
+		}
+
+	});
+
+}
+
+}
+```
+## Motivation
+
+Before this proposal, dynamic component creation was done using the `lwc:dynamic` directive. This directive required a component author to know component configuration details like the property and event names at design time which breaks down for use cases that need true dynamic component creation. For instance, in the case of property values or event handlers one would have to keep a reference to the element and set the information retrieved at runtime manually.
+
+```js
+element;
+
+@api
+moduleName;
+
+connectedCallback() {
+	const ctor = await import(this.moduleName);
+	this.element = createComponent('c-lazy-cmp', { is: ctor });
+	// element must be manually inserted into the dom
+	this.template.querySelector('div').appendChild(this.element)
+}
+
+@api
+
+set properties(newProperties) {
+
+this.props = newProperties;
+
+Object.assign(this.element, this.props);
+
+}
+
+get attributes() { return this.props; }
+
+```
+A new tool must be introduced that will allow component authors to configure their components with information retrieved at runtime without relying on maintaining element references or manually manipulating the DOM.
+
+## Proposal
```

Comment:
It would be great to add a section about how other frameworks are approaching this issue.